app/src/lib/git/worktree.ts

```diff
@@ -0,0 +1,143 @@
+import * as Os from 'os'
+import * as Path from 'path'
+import * as FSE from 'fs-extra'
+
+import { git } from './core'
+
+import { Repository, LinkedWorkTree } from '../../models/repository'
+import { getMatches } from '../helpers/regex'
+
+/** Enumerate the list of work trees reported by Git for a repository */
+export async function listWorkTrees(
+  repository: Repository
+): Promise<ReadonlyArray<LinkedWorkTree>> {
+  const result = await git(
+    ['worktree', 'list', '--porcelain'],
+    repository.path,
+    'listWorkTrees'
+  )
+
+  const worktrees = new Array<LinkedWorkTree>()
+
+  // the porcelain output from git-worktree covers multiple lines
+  const listWorkTreeRe = /worktree (.*)\nHEAD ([a-f0-9]*)\n(branch .*|detached)\n/gm
+
+  getMatches(result.stdout, listWorkTreeRe).forEach(m => {
+    if (m.length === 4) {
+      worktrees.push({
+        path: m[1],
+        head: m[2],
+      })
+    } else {
+      log.debug(
+        `[listWorkTrees] match '${
+          m[0]
+        }' does not have the expected data or output. Skipping...`
+      )
+    }
+  })
+
+  return worktrees
+}
+
+/** Create a new work tree at the desired location on disk */
+async function addWorkTree(
+  repository: Repository,
+  path: string,
+  commit: string
+): Promise<LinkedWorkTree> {
+  await git(
+    ['worktree', 'add', '-f', path, commit],
+    repository.path,
+    'addWorkTree'
+  )
+
+  // Because Git doesn't give enough information from stdout for the previous
+  // Git call, this function enumerates the available worktrees to find the
+  // expected worktree
+
+  const workTrees = await listWorkTrees(repository)
+
+  const directoryName = Path.basename(path)
+  const workTree = workTrees.find(t => Path.basename(t.path) === directoryName)
+
+  if (workTree == null) {
+    throw new Error(
+      `[addWorkTree] Unable to find created worktree at path ${path}`
+    )
+  }
+
+  return workTree
+}
+
+/** Cleanup the temporary worktree at a given location */
+export async function removeWorkTree(
+  repository: Repository,
+  path: string
+): Promise<true> {
+  await git(
```

Comment:
If the worktree is not able to be removed, is returning `true` still cool?

Reply:
if any errors are occurred dugite should throw an error, in which case the caller would not get `true` returned so i think this is ok. does that sound right?